include/swift/SILOptimizer/Utils/CanonicalOSSALifetime.h

```diff
@@ -0,0 +1,187 @@
+/// scope and mark_dependence is associated with an end_dependence,
+/// canonicalization will work everywhere as intended.
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_SILOPTIMIZER_UTILS_CANONICALOSSALIFETIME_H
+#define SWIFT_SILOPTIMIZER_UTILS_CANONICALOSSALIFETIME_H
+
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/SetVector.h"
+#include "swift/SIL/SILInstruction.h"
+#include "swift/SILOptimizer/Utils/PrunedLiveness.h"
+
+namespace swift {
+
+struct CanonicalOSSALifetime;
+
+/// Top-Level API: rewrites copies and destroys within \p def's extended
+/// lifetime. \p lifetime caches transient analysis state across multiple calls
+/// and indicates whether any SILAnalyses must be invalidated.
+///
+/// Return false if the OSSA structure cannot be recognized (with a proper OSSA
+/// representation this will always return true).
+bool canonicalizeValueLifetime(SILValue def, CanonicalOSSALifetime &lifetime);
+
+/// Find the original definition of a potentially copied value.
+///
+/// This use-def walk must be consistent with the def-use walks performed within
+/// the CanonicalOSSALifetime implementation.
+inline SILValue getCanonicalCopiedDef(SILValue v) {
+  while (true) {
+    if (auto *copy = dyn_cast<CopyValueInst>(v)) {
+      v = copy->getOperand();
+      continue;
+    }
+    return v;
+  }
+}
+
+/// Information about consumes on the extended-lifetime boundary. Consuming uses
+/// within the lifetime are not included--they will consume a copy after
+/// rewriting.
+///
+/// This result remains valid during copy rewriting. The only instructions
+/// referenced it contains are consumes that cannot be deleted.
+class CanonicalOSSAConsumeInfo {
+  // Map blocks on the lifetime boundary to the last consuming instruction.
+  llvm::SmallDenseMap<SILBasicBlock *, SILInstruction *, 4> finalBlockConsumes;
+
+public:
+  bool empty() const { return finalBlockConsumes.empty(); }
+
+  void clear() { finalBlockConsumes.clear(); }
+
+  void recordFinalConsume(SILInstruction *inst) {
+    assert(!finalBlockConsumes.count(inst->getParent()));
+    finalBlockConsumes[inst->getParent()] = inst;
+  }
+
+  // Return true if this instruction is marked as a final consume point of the
+  // current def's live range. A consuming instruction can only be claimed once
+  // because instructions like `tuple` can consume the same value via multiple
+  // operands.
+  bool claimConsume(SILInstruction *inst) {
+    auto destroyPos = finalBlockConsumes.find(inst->getParent());
+    if (destroyPos != finalBlockConsumes.end() && destroyPos->second == inst) {
+      finalBlockConsumes.erase(destroyPos);
+      return true;
+    }
+    return false;
+  }
+
+  SWIFT_ASSERT_ONLY_DECL(void dump() const LLVM_ATTRIBUTE_USED);
+};
+
+/// Canonicalize OSSA lifetimes.
+///
+/// Allows the allocation of analysis state to be reused across calls to
+/// canonicalizeValueLifetime().
+struct CanonicalOSSALifetime {
```

Comment:
This name makes it sound like this is something related to the actual lifetime itself instead of the state used for Canonicalizing OSSA lifetimes. Why not something like: ``OSSALifetimeCanonicalizer``?

Reply:
This just represents the canonical OSSA lifetime before SIL transformation. That's all it is. It doesn't do anything.